File: Auto/Test_code/C++/C++ first edition/C_Joy.cpp
#include <iostream>
#include <SDL2/SDL.h>
#include "MotorModule_new.h"

int main() {
    if (SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_GAMECONTROLLER) != 0) {
        std::cout << "Failed to initialize SDL: " << SDL_GetError() << std::endl;
        return 1;
    }

    SDL_GameController* controller = nullptr;
    for (int i = 0; i < SDL_NumJoysticks(); i++) {
        if (SDL_IsGameController(i)) {
            controller = SDL_GameControllerOpen(i);
            if (controller != nullptr) {
                break;
            } else {
                std::cout << "Failed to open game controller: " << SDL_GetError() << std::endl;
                return 1;
            }
        }
    }

    Motor motor(2, 3, 4, 17, 22, 27, 14, 15, 18);
    float prevX = SDL_JoystickGetAxis(SDL_GameControllerGetJoystick(controller), 0) / 32767.0f;
    float prevY = SDL_JoystickGetAxis(SDL_GameControllerGetJoystick(controller), 1) / 32767.0f;

    // Main Loop
    while (true) {
        SDL_Event event;
        while (SDL_PollEvent(&event)) {
            if (event.type == SDL_JOYBUTTONDOWN) {
                if (event.jbutton.button == 0) {
                    std::cout << "button 0 down" << std::endl;
                    motor.move(0.6, 0, 0.1);
                } else if (event.jbutton.button == 1) {
                    std::cout << "button 1 down" << std::endl;
                    motor.move(-0.6, 0, 0.1);
                } else if (event.jbutton.button == 2) {
                    std::cout << "button 2 down" << std::endl;
                    motor.move(0, -0.8, 0.1);
                } else if (event.jbutton.button == 3) {
                    std::cout << "button 3 down" << std::endl;
                    motor.move(0, 0.8, 0.1);
                } else if (event.jbutton.button == 5) {
                    std::cout << "button 5 down" << std::endl;
                } else if (event.jbutton.button == 6) {
                    std::cout << "button 6 down" << std::endl;
                } else if (event.jbutton.button == 7) {
                    std::cout << "button 7 down" << std::endl;
                } else if (event.jbutton.button == 8) {
                    std::cout << "button 8 down" << std::endl;
                }
            }
        }

        float x = SDL_JoystickGetAxis(SDL_GameControllerGetJoystick(controller), 0) / 32767.0f;
        float y = SDL_JoystickGetAxis(SDL_GameControllerGetJoystick(controller), 1) / 32767.0f;
        if (x != prevX || y != prevY) {
            std::cout << "Joystick values: x=" << x << ", y=" << y << std::endl;
            motor.move(y * 0.8, x * 0.6, 0.1);
            prevX = x;
            prevY = y;
        }
    }

    SDL_GameControllerClose(controller);
    SDL_Quit();

    return 0;
}

File: Auto/final_code/python/C_Joy.py
import pygame
from MotorModule_new import Motor

pygame.init()

# Initialize the joystick
joystick = pygame.joystick.Joystick(0)
joystick.init()

# Xbox Joystick Axis:
# Axis 0 up down, down value is -1, up value is 1
# Axis 1 Left, Right, Left value is: -1, right value is 1
# center is always 0

##################
motor = Motor(2, 3, 4, 17, 22, 27, 14, 15, 18)

# Previous joystick values
prev_x, prev_y = joystick.get_axis(0), joystick.get_axis(1)

# Main Loop
while True:
    # Check for joystick events
    for event in pygame.event.get():
        if event.type == pygame.JOYBUTTONDOWN:
            if event.button == 0:
                print("button 0 down")
                motor.move(0.6, 0, 0.1)
            elif event.button == 1:
                print("button 1 down")
                motor.move(-0.6, 0, 0.1)
            elif event.button == 2:
                print("button 2 down")
                motor.move(0, -0.8, 0.1)
            elif event.button == 3:
                print("button 3 down")
                motor.move(0, 0.8, 0.1)
            elif event.button == 5:
                print("button 5 down")
            elif event.button == 6:
                print("button 6 down")
            elif event.button == 7:
                print("button 7 down")
            elif event.button == 8:
                print("button 8 down")

    # Check the joystick's axis values
    x, y = joystick.get_axis(0), joystick.get_axis(1)
    if x != prev_x or y != prev_y:
        print(f"Joystick values: x={x}, y={y}")
        # Move the motors according to the joystick's axis values
        motor.move(y * 0.8, x * 0.6, 0.1)

        # Update previous joystick values
        prev_x, prev_y = x, y

